File: LCA.java
package HW.HW9;


public class LCA {

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

        if(root == null){
            return null;
        }
        
        if( root == p || root == q){
            return root;

        }

        TreeNode leftTest = lowestCommonAncestor(root.left, p, q);
        TreeNode rightTest = lowestCommonAncestor(root.right, p, q);

        if(!(rightTest == null) && !(leftTest == null)){
            return root;
        }

        else{

            if(rightTest == null){
                return leftTest;
            }
            return rightTest;
        }
    

    }
    public static void main(String[] args) {

        TreeNode root = new TreeNode(4);
        root.right = new TreeNode(8);
        root.left = new TreeNode(3);
        root.left.left = new TreeNode(1);
        
        root.right.left = new TreeNode(5);
     
        root.right.right = new TreeNode(9);

        
        LCA test = new LCA();

        TreeNode result = test.lowestCommonAncestor(root, root.left, root.left.left);

        System.out.println(result.val);
        
    }

    

}

File: LCA.py

class LCA:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root is None:
            return None
        if root == p or root == q:
            return root
        
        rightTest = self.lowestCommonAncestor(root.right, p, q)
        leftTest = self.lowestCommonAncestor(root.left, p, q)

        if rightTest and leftTest:
            return root
        else:
            return rightTest or leftTest

class TreeNode(object):

     def __init__(self, val=0, left=None, right=None):

         self.val = val

         self.left = left

         self.right = right

LCA = LCA()
root = TreeNode(4)
root.right = TreeNode(8)
root.left = TreeNode(3)
root.left.left = TreeNode(1)
root.right.left = TreeNode(5)
root.right.right = TreeNode(9)

result = LCA.lowestCommonAncestor(root, root.left, root.left.left)

print(result.val)